### PR TITLE
Handle Netty channels returning null for remoteAddress

### PR DESCRIPTION
**Pull request by robbavey**
Handles the case where the Netty channel returns remoteAddress
as null, without spamming the logs with warning messages for each
instance.

Also, changed the the method to retrieve the @metadata field from
the the messageData hash from get to '[]' - this was silently
failing in tests.

Mitigates #269

### CODE REVIEW

**Comment by jordansissel** [comment]
> where the Netty channel returns remoteAddress as null

The description sounds like we are working around a bug in Netty. Why is `remoteAddress` null? More generally, how would we have a connection with no remote address?

**Comment by robbavey** [comment]
@jordansissel There appear to be a couple of places in the Netty code where remoteAddress can return null - 
https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/AbstractChannel.java#L197 and 

https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/AbstractServerChannel.java

The comment here https://github.com/logstash-plugins/logstash-input-beats/issues/269#issuecomment-346412968 seems to point the finger at the return value from channel(). 

I tried repro'ing the issue locally, and on aws and alicloud, but to no avail...

**Comment by jordansissel** [comment]
@robbavey weird; it returns null when an exception is thrown from the underlying implementation fetching the SocketAddress. I think this only happens when the socket is closed? I wonder how to safely resolve this -- we want to include the peer address in the events, and when we receive the event, the socket is open.

**Comment by jordansissel** [comment]
@robbavey I'm not stating this well, having trouble with words today...

Users are reporting that this warning is "spamming" their log messages. This indicates the message happens *constantly* and the problem of a null socketAddress is frequent.  I would like us to understand why this happens now, and figure out a solution to resolve it.

If we solve it by debug logging but otherwise not warning, then these users' events will have no `ip_address` field and the next step for users is to notice that `ip_address` is missing, then file a bug about `ip_address` being missing. I feel that this PR just relocates the problem instead of fixing the underlying problem (socketAddress is null when it should not be null)

Maybe we need to cache the socketAddress when the socket is first connected? I don't know how these problems are happening, so it's hard to brainstorm a solution.

**Comment by robbavey** [comment]
@jordansissel - you're right. Let me look into trying to find the root cause or find a better way to mitigate the symptoms when it happens, rather than just muting the problem. It's proving an interesting one to track down...

**Comment by robbavey** [comment]
@jordansissel I've updated the code to populate the ip address on connection, rather than for every message.

cc @ph

**Comment by robbavey** [comment]
ping @jordansissel (Pretty sure the build issue is not related to this PR)

**Comment by robbavey** [comment]
Rebased after #275, #278

**Comment by jordansissel** [comment]
Thanks; will review today!

File: lib/logstash/inputs/beats/message_listener.rb [hunk 1]
```diff
@@ -83,8 +78,29 @@ def codec(ctx)
...
+        remote_address = ctx.channel.remoteAddress
+        # Netty allows remoteAddress to be nil, which can cause a lot of log entries - see
+        # https://github.com/logstash-plugins/logstash-input-beats/issues/269
+        if remote_address.nil?
+          input.logger.debug("Cannot retrieve remote IP address for beats input - remoteAddress is nil")
+          return nil
+        end
+        remote_address.getAddress.getHostAddress
+      rescue => e # This should not happen, but should not block the beats input
+        input.logger.warn("Could not retrieve remote IP address for beats input.", :message => e)
```
Comment (jordansissel) [inline lib/logstash/inputs/beats/message_listener.rb]: change `:message` to `:reason` or `:error` maybe?

**Pull request closed without merging**